src/ast/semantic_analyser.cpp

```diff
@@ -226,16 +226,16 @@ void SemanticAnalyser::visit(Call &call)
   else if (call.func == "str") {
     if (check_varargs(call, 1, 2)) {
       check_arg(call, Type::integer, 0);
+      uint64_t strlen = bpftrace_.strlen_;
+      call.type = SizedType(Type::string, strlen);
```

Comment:
@Birch-san do you see any downside on moving these assignments up here?

Reply:
@mmarchini I think we'll get away with it.

The "str" codegen doesn't actually look at `call.type`. Instead:

- It allocates `bpftrace_.strlen_` bytes for the string buffer
- It consults `call.vargs->at(1)` to decide the size of the `probe_read_str()` read

So, we could delete this section as well:

```c++
auto &strlen_arg = *call.vargs->at(1);
if (strlen_arg.is_literal) {
  strlen = static_cast<Integer&>(strlen_arg).n;
}
```

This code is just an artefact of when I tried to support parameterized allocation of string buffers. I aborted that because of how hard it was to make printf() support arguments whose lengths were decided at runtime. See https://github.com/iovisor/bpftrace/issues/305 _("We may find that we want a dynamically-sized perf event output structure")_.

We may revisit this when we undertake [support for big strings](https://github.com/iovisor/bpftrace/issues/305), but even then it's just a performance optimization (allocating smallest string buffer possible, when input size is known).